File: Sai Nitheesh/quadtree.c
#include <stdio.h>
#include <stdlib.h>

#define CAPACITY 2

// Structure for a point
typedef struct Point {
    int x, y;
} Point;

// Structure for QuadTree node
typedef struct QuadTree {
    int x, y, w, h;   // boundary (rectangle)
    Point points[CAPACITY];
    int count;

    struct QuadTree *nw, *ne, *sw, *se;
    int divided;
} QuadTree;

// Create a new QuadTree node
QuadTree* createNode(int x, int y, int w, int h) {
    QuadTree* node = (QuadTree*)malloc(sizeof(QuadTree));
    node->x = x;
    node->y = y;
    node->w = w;
    node->h = h;
    node->count = 0;
    node->divided = 0;
    node->nw = node->ne = node->sw = node->se = NULL;
    return node;
}

// Check if point is inside boundary
int contains(QuadTree* qt, Point p) {
    return (p.x >= qt->x && p.x < qt->x + qt->w &&
            p.y >= qt->y && p.y < qt->y + qt->h);
}

// Subdivide node into 4 children
void subdivide(QuadTree* qt) {
    int x = qt->x;
    int y = qt->y;
    int w = qt->w / 2;
    int h = qt->h / 2;

    qt->nw = createNode(x, y, w, h);
    qt->ne = createNode(x + w, y, w, h);
    qt->sw = createNode(x, y + h, w, h);
    qt->se = createNode(x + w, y + h, w, h);

    qt->divided = 1;
}

// Insert a point into QuadTree
void insert(QuadTree* qt, Point p) {
    if (!contains(qt, p))
        return;

    if (qt->count < CAPACITY) {
        qt->points[qt->count++] = p;
    } else {
        if (!qt->divided)
            subdivide(qt);

        insert(qt->nw, p);
        insert(qt->ne, p);
        insert(qt->sw, p);
        insert(qt->se, p);
    }
}

// Print points (for testing)
void printTree(QuadTree* qt) {
    for (int i = 0; i < qt->count; i++) {
        printf("(%d, %d)\n", qt->points[i].x, qt->points[i].y);
    }

    if (qt->divided) {
        printTree(qt->nw);
        printTree(qt->ne);
        printTree(qt->sw);
        printTree(qt->se);
    }
}

// Main function
int main() {
    QuadTree* qt = createNode(0, 0, 100, 100);

    Point p1 = {10, 20};
    Point p2 = {30, 40};
    Point p3 = {70, 80};
    Point p4 = {90, 10};

    insert(qt, p1);
    insert(qt, p2);
    insert(qt, p3);
    insert(qt, p4);

    printf("Stored Points:\n");
    printTree(qt);

    return 0;
}

File: karanjith/Quadtree.py
class Point:
    def __init__(self, x, y):
        self.x = x
        self.y = y

    def __repr__(self):
        return f"({self.x}, {self.y})"

class Boundary:
    def __init__(self, x, y, width, height):
        self.x = x
        self.y = y
        self.width = width
        self.height = height

    def contains(self, point):
        return (self.x - self.width <= point.x < self.x + self.width and
                self.y - self.height <= point.y < self.y + self.height)

    def intersects(self, range):
        return not (range.x - range.width > self.x + self.width or
                    range.x + range.width < self.x - self.width or
                    range.y - range.height > self.y + self.height or
                    range.y + range.height < self.y - self.height)

class Quadtree:
    def __init__(self, boundary, capacity):
        self.boundary = boundary
        self.capacity = capacity
        self.points = []
        self.divided = False

    def subdivide(self):
        x = self.boundary.x
        y = self.boundary.y
        w = self.boundary.width
        h = self.boundary.height

        ne = Boundary(x + w / 2, y - h / 2, w / 2, h / 2)
        nw = Boundary(x - w / 2, y - h / 2, w / 2, h / 2)
        se = Boundary(x + w / 2, y + h / 2, w / 2, h / 2)
        sw = Boundary(x - w / 2, y + h / 2, w / 2, h / 2)

        self.northeast = Quadtree(ne, self.capacity)
        self.northwest = Quadtree(nw, self.capacity)
        self.southeast = Quadtree(se, self.capacity)
        self.southwest = Quadtree(sw, self.capacity)
        self.divided = True

    def insert(self, point):
        if not self.boundary.contains(point):
            return False

        if len(self.points) < self.capacity:
            self.points.append(point)
            return True
        else:
            if not self.divided:
                self.subdivide()

            if self.northeast.insert(point): return True
            if self.northwest.insert(point): return True
            if self.southeast.insert(point): return True
            if self.southwest.insert(point): return True

        return False

    def query(self, range, found_points=None):
        if found_points is None:
            found_points = []

        if not self.boundary.intersects(range):
            return found_points

        for point in self.points:
            if range.contains(point):
                found_points.append(point)

        if self.divided:
            self.northwest.query(range, found_points)
            self.northeast.query(range, found_points)
            self.southwest.query(range, found_points)
            self.southeast.query(range, found_points)

        return found_points

if __name__ == "__main__":
    boundary = Boundary(200, 200, 200, 200)
    qt = Quadtree(boundary, 4)
    qt.insert(Point(50, 50))
    qt.insert(Point(150, 150))
    qt.insert(Point(250, 250))
    qt.insert(Point(350, 350))
    print(qt.query(Boundary(200, 200, 200, 200)))
